src/client/client_helper.ml

```diff
@@ -36,18 +36,29 @@ let default_create_client_context ~ca_cert ~creds ~protocol =
 
 
 let with_connection ~tls ~tcp_keepalive sa do_it = match tls with
-  | None -> Lwt_io.with_connection sa do_it
+  | None -> 
+     let fd = Lwt_unix.socket (Unix.domain_of_sockaddr sa) Unix.SOCK_STREAM 0 in
+     Tcp_keepalive.apply fd tcp_keepalive;
+     Lwt_unix.setsockopt fd Lwt_unix.TCP_NODELAY true;
+     Lwt_unix.set_close_on_exec fd;
+     Lwt_unix.connect fd sa >>= fun () ->
```

Comment:
Connecting could fail, so I think it should be put in the finalizer below, so we don't leak file descriptors in that case.
Actually it may be best to start the Lwt.finalize immediately after creating the fd, then we don't have to know/assume that the other functions never fail.

The first part of the ssl and non-ssl case are the same, so it looks like the code could be shared.